src/mlpack/core/cereal/pointer_variant_wrapper.hpp

```diff
@@ -0,0 +1,88 @@
+/**
+ * @file core/cereal/pointer_variant_wrapper.hpp
+ * @author Omar Shrit
+ *
+ * Implementation of a boost::variant wrapper to enable the serialization of
+ * the pointers inside boost variant in cereal
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
+#define MLPACK_CORE_CEREAL_POINTER_VARIANT_WRAPPER_HPP
+
+#include <cereal/archives/json.hpp>
+#include <cereal/archives/portable_binary.hpp>
+#include <cereal/archives/xml.hpp>
+#include <cereal/types/boost_variant.hpp>
+
+#include <boost/variant.hpp>
+
+#include "pointer_wrapper.hpp"
+
+namespace cereal {
+
+struct save_vistor : public boost::static_visitor<>
+{
+  template<class Archive, class T>
+  void operator()(Archive& ar, T t) const 
+  {
+    ar & CEREAL_POINTER(t);
+  }
+};
```

Comment:
I believe this might need to be something more like this:

```
template<typename Archive>
struct save_visitor : public boost::static_visitor<void>
{
  template<typename T>
  void operator()(T& t)
  {
    ar(CEREAL_POINTER(t));
  }
};
```

Don't forget the `void` template type to `static_visitor` like @robertohueso pointed out, and don't forget to use a reference for `t` in the `operator()` so that we don't copy the `T`.

You'll probably have to adapt the `load_visitor` the same way.  Honestly, since the code is the same in both, you might be able to just do `serialize_visitor` and use thise in `pointer_variant_wrapper`.

Reply:
In this case, the `ar` will be undefined if I remove the `Archive& ar` from the signature of `operator()`.

Reply:
Sorry I wrote this totally wrong :facepalm: let me try again:

```
template<typename Archive>
struct save_visitor : public boost::static_visitor<void>
{
  save_visitor(Archive& ar) : ar(ar) { }

  template<typename T>
  void operator()(T& t)
  {
    ar(CEREAL_POINTER(t));
  }

  Archive& ar;
};
```